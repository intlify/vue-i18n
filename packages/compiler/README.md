# @intlify/message-compiler

## :copyright: License

[MIT](http://opensource.org/licenses/MIT)
