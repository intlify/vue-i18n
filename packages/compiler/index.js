'use strict'

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./dist/compiler.cjs.prod.js')
} else {
  module.exports = require('./dist/compiler.cjs.js')
}
