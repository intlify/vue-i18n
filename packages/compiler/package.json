{
  "name": "@intlify/compiler",
  "version": "9.0.0",
  "description": "@intlify/compiler",
  "main": "index.js",
  "module": "dist/compiler.esm-bundler.js",
  "types": "dist/compiler.d.ts",
  "unpkg": "dist/compiler.global.js",
  "jsdelivr": "dist/compiler.global.js",
  "files": [
    "index.js",
    "dist"
  ],
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/intlify/vue-i18n-next.git",
    "directory": "packages/compiler"
  },
  "buildOptions": {
    "name": "IntlifyCompiler",
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "keywords": [
    "intlify",
    "i18n",
    "message-format",
    "compiler"
  ],
  "author": {
    "name": "kazuya kawaguchi",
    "email": "kawakazu80@gmail.com"
  },
  "browser": "dist/compiler.esm-browser.js",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/intlify/vue-i18n-next/issues"
  },
  "homepage": "https://github.com/intlify/vue-i18n-next/tree/master/packages/compiler#readme",
  "dependencies": {
    "@intlify/shared": "9.0.0"
  }
}
