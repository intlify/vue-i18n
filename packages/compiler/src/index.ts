import { isObject } from '@intlify/shared'

export type Foo = 'Foo'

console.log('hello compiler!', isObject({}))
