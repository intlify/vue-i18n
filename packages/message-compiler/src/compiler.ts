import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
