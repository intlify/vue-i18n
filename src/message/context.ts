import {
  isNumber,
  isFunction,
  toDisplayString,
  isObject,
  isString,
  isPlainObject
} from '../utils'

export type PluralizationRule = (
  choice: number,
  choicesLength: number,
  orgRule?: PluralizationRule
) => number

export type PluralizationRules = { [locale: string]: PluralizationRule }
export type LinkedModify = (value: unknown, type: string) => unknown
export type LinkedModifiers = { [key: string]: LinkedModify }
export type MessageFunction = (ctx: MessageContext) => unknown
export type MessageFucntions = { [key: string]: MessageFunction }
export type MessageResolveFunction = (key: string) => MessageFunction
export type NamedValue<T = {}> = T & { [prop: string]: unknown }
export type MessageNormalize = (values: unknown[]) => unknown
export type MessageInterpolate = (val: unknown) => unknown

export interface MessageProcessor {
  type?: string
  interpolate?: MessageInterpolate
  normalize?: MessageNormalize
}

export type MessageContextOptions<N = {}> = {
  parent?: MessageContext
  locale?: string
  list?: unknown[]
  named?: NamedValue<N>
  modifiers?: LinkedModifiers
  pluralIndex?: number
  pluralRules?: PluralizationRules
  messages?: MessageFucntions | MessageResolveFunction // TODO: need to design resolve message function?
  processor?: MessageProcessor
}

export type MessageContext = {
  list: (index: number) => unknown
  named: (key: string) => unknown
  pluralIndex: number
  pluralRule: PluralizationRule
  orgPluralRule?: PluralizationRule
  modifier: (name: string) => LinkedModify
  message: (name: string) => MessageFunction
  type: string
  interpolate: MessageInterpolate
  normalize: MessageNormalize
}

const DEFAULT_MODIFIER = (str: unknown): unknown => str
const DEFAULT_MESSAGE = (ctx: MessageContext): unknown => '' // eslint-disable-line
export const DEFAULT_MESSAGE_DATA_TYPE = 'text'
const DEFAULT_NORMALIZE = (values: unknown[]): unknown =>
  values.length === 0 ? '' : values.join('')
const DEFAULT_INTERPOLATE = toDisplayString

function pluralDefault(choice: number, choicesLength: number): number {
  choice = Math.abs(choice)
  if (choicesLength === 2) {
    // prettier-ignore
    return choice
      ? choice > 1
        ? 1
        : 0
      : 1
  }
  return choice ? Math.min(choice, 2) : 0
}

function getPluralIndex(options: MessageContextOptions): number {
  // prettier-ignore
  const index = isNumber(options.pluralIndex)
    ? options.pluralIndex
    : -1
  // prettier-ignore
  return options.named && (isNumber(options.named.count) || isNumber(options.named.n))
    ? isNumber(options.named.count)
      ? options.named.count
      : isNumber(options.named.n)
        ? options.named.n
        : index
    : index
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function normalizeNamed(pluralIndex: number, named: any): void {
  if (!named.count) {
    named.count = pluralIndex
  }
  if (!named.n) {
    named.n = pluralIndex
  }
}

export function createMessageContext<N = {}>(
  options: MessageContextOptions<N> = {}
): MessageContext {
  const locale = options.locale
  // TODO: should be implemented warning message
  const pluralIndex = getPluralIndex(options)

  // TODO: should be implemented warning message
  const pluralRule =
    isObject(options.pluralRules) &&
    isString(locale) &&
    isFunction(options.pluralRules[locale])
      ? options.pluralRules[locale]
      : pluralDefault
  const orgPluralRule =
    isObject(options.pluralRules) &&
    isString(locale) &&
    isFunction(options.pluralRules[locale])
      ? pluralDefault
      : undefined

  const _list = options.list || []
  // TODO: should be implemented warning message
  const list = (index: number): unknown => _list[index]

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const _named = options.named || ({} as any)
  isNumber(options.pluralIndex) && normalizeNamed(pluralIndex, _named)
  // TODO: should be implemented warning message
  const named = (key: string): unknown => _named[key]

  // TODO: should be implemented warning message
  const modifier = (name: string): LinkedModify =>
    options.modifiers ? options.modifiers[name] : DEFAULT_MODIFIER

  const message = (name: string): MessageFunction => {
    // TODO: need to design resolve message function?
    // prettier-ignore
    const msg = isFunction(options.messages)
      ? options.messages(name)
      : isObject(options.messages)
        ? options.messages[name]
        : false
    return !msg
      ? options.parent
        ? options.parent.message(name) // resolve from parent messages
        : DEFAULT_MESSAGE
      : msg
  }

  const type =
    isPlainObject(options.processor) && isString(options.processor.type)
      ? options.processor.type
      : DEFAULT_MESSAGE_DATA_TYPE

  const normalize =
    isPlainObject(options.processor) && isFunction(options.processor.normalize)
      ? options.processor.normalize
      : DEFAULT_NORMALIZE

  const interpolate =
    isPlainObject(options.processor) &&
    isFunction(options.processor.interpolate)
      ? options.processor.interpolate
      : DEFAULT_INTERPOLATE

  return {
    list,
    named,
    pluralIndex,
    pluralRule,
    orgPluralRule,
    modifier,
    message,
    type,
    interpolate,
    normalize
  }
}
