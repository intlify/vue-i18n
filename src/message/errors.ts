import { SourceLocation } from './location'
import { format } from '../utils'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: number
  domain?: CompileDomain
  location?: SourceLocation
}

export interface CreateCompileErrorOptions {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,
  UNEXPECTED_LEXICAL_ANALYSIS,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`,
  [CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS]: `Unexpected lexical analysis in token: '{0}'`
}

export function createCompileError<T extends number>(
  code: T,
  loc: SourceLocation | null,
  optinos: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = optinos
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  if (loc) {
    error.location = loc
  }
  error.domain = domain
  return error
}

export function defaultOnError(error: CompileError): never {
  throw error
}
