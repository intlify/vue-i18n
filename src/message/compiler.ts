import { CompileOptions } from './options'
import { ResourceNode, createParser } from './parser'
import { transform } from './transformer'
import { generate } from './generator'
import { CompileError, defaultOnError } from './errors'
import { MessageFunction, MessageFunctions } from './runtime'
import { warn, format, isBoolean } from '../utils'

export interface CompileResult {
  code: string
  ast: ResourceNode
  // TODO: should be implemetend sourcemap
  // map?: RawSourceMap
}

const RE_HTML_TAG = /<\/?[\w\s="/.':;#-\/]+>/
const WARN_MESSAGE = `Detected HTML in '{source}' message. Recommend not using HTML messages to avoid XSS.`

function checkHtmlMessage(source: string, options: CompileOptions): void {
  const warnHtmlMessage = isBoolean(options.warnHtmlMessage)
    ? options.warnHtmlMessage
    : true
  if (warnHtmlMessage && RE_HTML_TAG.test(source)) {
    warn(format(WARN_MESSAGE, { source }))
  }
}

const defaultOnCacheKey = (source: string): string => source
let compileCache: unknown = Object.create(null)

export function clearCompileCache(): void {
  compileCache = Object.create(null)
}

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CompileResult {
  // parse source codes
  const parser = createParser({ ...options })
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, { ...options })

  // generate javascript codes
  const code = generate(ast, { ...options })

  return { ast, code }
}

export function compile<T = string>(
  source: string,
  options: CompileOptions = {}
): MessageFunction<T> {
  // check HTML message
  __DEV__ && checkHtmlMessage(source, options)

  // check caches
  const onCacheKey = options.onCacheKey || defaultOnCacheKey
  const key = onCacheKey(source)
  const cached = (compileCache as MessageFunctions<T>)[key]
  if (cached) {
    return cached
  }

  // compile error detecting
  let occured = false
  const onError = options.onError || defaultOnError
  options.onError = (err: CompileError): void => {
    occured = true
    onError(err)
  }

  // compile
  const { code } = baseCompile(source, options)

  // evaluate function
  const msg = new Function(`return ${code}`)() as MessageFunction<T>

  // if occured compile error, don't cache
  return !occured ? ((compileCache as MessageFunctions<T>)[key] = msg) : msg
}
