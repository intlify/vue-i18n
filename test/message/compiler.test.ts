import { compile } from '../../src/message/compiler'

/* eslint-disable no-irregular-whitespace */
test(`@.caml:{'no apples'} | {0} apple | {n}　apples`, () => {
  const code = compile(`@.caml:{'no apples'} | {0} apple | {n}　apples`)
  expect(code.toString()).toMatchSnapshot('code')
})
/* eslint-enable no-irregular-whitespace */

describe('edge cases', () => {
  test(` | | | `, () => {
    const code = compile(` | | | `, {
      onError(error) {
        expect({ ...error, message: error.message }).toMatchSnapshot('error')
      }
    })
    expect(code.toString()).toMatchSnapshot('code')
  })
})
