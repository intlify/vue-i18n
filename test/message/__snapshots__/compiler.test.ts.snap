// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@.caml:{'no apples'} | {0} apple | {n}　apples: code 1`] = `
"function __msg__ (ctx) {
  return [
    ctx.normalize([
      ctx.modifier(\\"caml\\")(ctx.message(\\"no apples\\")(ctx), ctx.type)
    ]), ctx.normalize([
      ctx.interpolate(ctx.list(0)), \\" apple\\"
    ]), ctx.normalize([
      ctx.interpolate(ctx.named(\\"n\\")), \\"　apples\\"
    ])
  ][ctx.pluralRule(ctx.pluralIndex, 3, ctx.orgPluralRule)]
}"
`;

exports[`edge cases  | | | : code 1`] = `
"function __msg__ (ctx) {
  return [
    ctx.normalize([
      
    ]), ctx.normalize([
      
    ]), ctx.normalize([
      
    ]), ctx.normalize([
      
    ])
  ][ctx.pluralRule(ctx.pluralIndex, 4, ctx.orgPluralRule)]
}"
`;

exports[`edge cases  | | | : error 1`] = `
Object {
  "code": 10,
  "domain": "parser",
  "location": Object {
    "end": Object {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "message": "Plural must have messages",
}
`;
